### ISSUE TITLE
Integration BYRON_ADDRESS tests fail intermittently

### ISSUE THREAD

**Issue by piotr-iohk** [cardano-foundation/cardano-wallet#1644]
# Context
 
https://buildkite.com/input-output-hk/cardano-wallet/builds/7409#d6d784e5-27e0-47ab-98ee-196d96c12868/6-7586

# Test Case

Those seems to be failing randomly:
 - ADDRESS_CREATE_04, ADDRESS_CREATE_06
 - CLI_ADDRESS_CREATE_04, CLI_ADDRESS_CREATE_06

**On [buildkite](https://buildkite.com/input-output-hk/cardano-wallet/builds/7409#d6d784e5-27e0-47ab-98ee-196d96c12868/6-7586):**
```

Failures:
--
  |  
  | test/integration/Test/Integration/Byron/Scenario/API/Addresses.hs:205:30:
  | 1) API Specifications, BYRON_ADDRESSES, ADDRESS_CREATE_04 - Can list address after creating it
  | uncaught exception: IOException of type UserError
  | user error (expectListField: trying to access the #0 element from a list but there's none! )
  |  
  | To rerun use: --match "/API Specifications/BYRON_ADDRESSES/ADDRESS_CREATE_04 - Can list address after creating it/"
  |  
  | test/integration/Test/Integration/Byron/Scenario/API/Addresses.hs:257:11:
  | 2) API Specifications, BYRON_ADDRESSES, ADDRESS_CREATE_06 - Cannot create an address that already exists
  | expected: Status {statusCode = 409, statusMessage = "Conflict"}
  | but got: Status {statusCode = 201, statusMessage = "Created"}

```

**Locally:**

```
stack test cardano-wallet-byron:integration --test-arguments "--match ADDRESS"
```

Most of the time it passes, but every once in a while I was getting those errors as above or from the equivalent CLI tests:

```
  test/integration/Test/Integration/Byron/Scenario/CLI/Addresses.hs:239:30: 
  1) CLI Specifications, BYRON_CLI_ADDRESSES, CLI_ADDRESS_CREATE_04 - Can list address after creating it
       uncaught exception: IOException of type UserError
       user error (expectCliListField: trying to access the #0 element from a list but there's none! )


  test/integration/Test/Integration/Byron/Scenario/CLI/Addresses.hs:290:5: 
  1) CLI Specifications, BYRON_CLI_ADDRESSES, CLI_ADDRESS_CREATE_06 - Cannot create an address that already exists
       "Please enter your passphrase: *****************\nOk.\n" does not contain "I already know of such address."

  To rerun use: --match "/CLI Specifications/BYRON_CLI_ADDRESSES/CLI_ADDRESS_CREATE_06 - Cannot create an address that already exists/"
```

# Failure / Counter-example

---

# Resolution 

See: https://github.com/input-output-hk/cardano-wallet/commit/73294aeb92297b890e0a67d9848dbdaf7c359fae

---

# QA

**Comment by piotr-iohk** [comment]
I think it is still an issue. The NB failed for the same reason -> https://buildkite.com/input-output-hk/cardano-wallet/builds/7450#ff424d99-21b8-4fdb-a2dd-7ffdbab779c6.

Also, I think I managed to observe it locally using latest cardano-wallet version (`2020.5.6 (git revision: ce8d18d4b8eabd4ee8635f6ef0ced60ac79f0037)`.

I have reproduced it using our docker-compose and mainnet:
```
NETWORK=mainnet docker-compose up
```
 - Create a random wallet
 - `POST
/byron-wallets/{walletId}/addresses` (while the wallet is still syncing)

and every now and again the address is not created despite the log says otherwise:

```
cardano-wallet_1  | [cardano-wallet.wallet-engine:Info:145612] [2020-05-12 09:10:12.63 UTC] f0e49a24: local tip: 629cf6dd-[1.12417#33999]
cardano-wallet_1  | [cardano-wallet.api-server:Info:145998] [2020-05-12 09:10:12.72 UTC] [RequestId 149] [POST] /v2/byron-wallets/f0e49a246e32511a3be82395dfd19ac55bdb3772/addresses
cardano-wallet_1  | [cardano-wallet.wallet-engine:Info:145612] [2020-05-12 09:10:12.72 UTC] f0e49a24: Applying blocks [1.12418 ... 1.13417]
cardano-wallet_1  | [cardano-wallet.api-server:Info:145998] [2020-05-12 09:10:12.79 UTC] [RequestId 149] 201 Created in 0.065642552s
cardano-wallet_1  | [cardano-wallet.api-server:Info:146006] [2020-05-12 09:10:12.79 UTC] [RequestId 150] [GET] /v2/byron-wallets/f0e49a246e32511a3be82395dfd19ac55bdb3772
cardano-wallet_1  | [cardano-wallet.wallet-engine:Info:145612] [2020-05-12 09:10:12.82 UTC] f0e49a24: Creating checkpoint at 7179731f-[1.13417#34999]
```

but doing: `GET
/byron-wallets/{walletId}/addresses` shows nothing.

Retrying `POST
/byron-wallets/{walletId}/addresses` eventually creates one.

The fact that there is no address being created perfectly explains the errors on integration tests:

```

Failures:
--
  |  
  | test/integration/Test/Integration/Byron/Scenario/API/Addresses.hs:226:30:
  | 1) API Specifications, BYRON_ADDRESSES, ADDRESS_CREATE_04 - Can list address after creating it
  | uncaught exception: IOException of type UserError
  | user error (expectListField: trying to access the #0 element from a list but there's none! )
  |  
  | To rerun use: --match "/API Specifications/BYRON_ADDRESSES/ADDRESS_CREATE_04 - Can list address after creating it/"
  |  
  | test/integration/Test/Integration/Byron/Scenario/API/Addresses.hs:278:11:
  | 2) API Specifications, BYRON_ADDRESSES, ADDRESS_CREATE_06 - Cannot create an address that already exists
  | expected: Status {statusCode = 409, statusMessage = "Conflict"}
  | but got: Status {statusCode = 201, statusMessage = "Created"}


```

**Comment by Anviking** [comment]
I think I managed to more consistently reproduce by running `stress --cpu 13 --timeout 600` at the same time. 

```
Results from repeatedly running locally
$ stack test cardano-wallet-byron:integration --test-arguments "--match ADDRESS"

on 16-threaded macOS machine

FAIL
SUCC
SUCC
FAIL
SUCC
SUCC
SUCC
FAIL
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC

$ stress --cpu 5 --timeout 600
SUCC
$ stress --cpu 7 --timeout 600
SUCC

$ stress --cpu 8 --timeout 600
FAIL

$ stress --cpu 15 --timeout 600
$ # Starting mainnet Daedalus in background

SUCC
SUCC
SUCC

$ stress --io 8 --timeout 600
FAIL
SUCC
SUCC

$ stress --io 16 --timeout 600
SUCC
SUCC

$ stress --io 30 --cpu 15 --timeout 600
SUCC

$ # Stopping mainnet daedalus in the background
$ stress --cpu 13 --timeout 600
FAIL
FAIL
SUCC
SUCC
FAIL
FAIL
FAIL
```

**Comment by Anviking** [comment]
With #1659 
```
Adding drastic fix making whole restoreBlocks atomic
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC
SUCC

```

**Comment by KtorZ** [comment]
I think this is now fixed by #1659 , good catch.

**Comment by piotr-iohk** [comment]
OK! Integration tests passing, not able to reproduce manually now.
